iothub/device/src/Pipeline/RetryDelegatingHandler.cs

```diff
@@ -621,170 +745,222 @@ private async Task RefreshSasTokenLoopAsync(DateTime refreshesOn, CancellationTo
-                    if (_deviceReceiveMessageEnabled)
+                    if (delay > TimeSpan.Zero)
                     {
-                        tasks.Add(base.EnableReceiveMessageAsync(cancellationToken));
+                        await Task.Delay(delay).ConfigureAwait(false);
                     }
 
-                    if (tasks.Any())
+                    // always reconnect.
+                    connectionStatusInfo = new ConnectionStatusInfo(ConnectionStatus.DisconnectedRetrying, ConnectionStatusChangeReason.CommunicationError);
+                    _onConnectionStatusChanged(connectionStatusInfo);
+                    CancellationToken cancellationToken = _handleDisconnectCts.Token;
+
+                    // This will recover to the status before the disconnect.
+                    await _internalRetryHandler.RunWithRetryAsync(async () =>
                     {
-                        await Task.WhenAll(tasks).ConfigureAwait(false);
-                    }
+                        if (Logging.IsEnabled)
+                            Logging.Info(this, "Attempting to recover subscriptions.", nameof(HandleDisconnectAsync));
 
-                    // Send the request for transport close notification.
-                    _transportClosedTask = HandleDisconnectAsync();
+                        await base.OpenAsync(cancellationToken).ConfigureAwait(false);
 
-                    _isOpen = true;
-                    connectionStatusInfo = new ConnectionStatusInfo(ConnectionStatus.Connected, ConnectionStatusChangeReason.ConnectionOk);
-                    _onConnectionStatusChanged(connectionStatusInfo);
+                        var tasks = new List<Task>(3);
+
+                        // This is to ensure that, if previously enabled, the callback to receive direct methods is recovered.
+                        if (_methodsEnabled)
+                        {
+                            tasks.Add(base.EnableMethodsAsync(cancellationToken));
+                        }
+
+                        // This is to ensure that, if previously enabled, the callback to receive twin properties is recovered.
+                        if (_twinEnabled)
+                        {
+                            tasks.Add(base.EnableTwinPatchAsync(cancellationToken));
+                        }
+
+                        // This is to ensure that, if previously enabled, the callback to receive C2D messages is recovered.
+                        if (_deviceReceiveMessageEnabled)
+                        {
+                            tasks.Add(base.EnableReceiveMessageAsync(cancellationToken));
+                        }
 
+                        if (tasks.Any())
+                        {
+                            await Task.WhenAll(tasks).ConfigureAwait(false);
+                        }
+
+                        // Send the request for transport close notification.
+                        _transportClosedTask = HandleDisconnectAsync();
```

Comment:
I'm worried about the lifetime of this handle disconnect code. It is called by someone else the first time and then by itself the second time.

Is any given instance ever pointing to the wrong token that it captures off of `_handleDisconnectCts`? Are we sure none of these calls duplicate over time?

I'd suggest diagraming it to be sure, and honestly this pattern is probably too fragile that anyone who touches this code and doesn't thoroughly understand this lifetime could break it.

I know you didn't introduce it in this PR, but perhaps we should look for a simpler impl like a callback.